### DISCUSSION TITLE
Object oriented programming

### DISCUSSION THREAD

**Discussion by naturalmechanics** [vlang/v#15167] [General]
Hi, I am a newcomer.
I hope this is the write place to ask. I dont see any mention of Object oriented. Is object oriented planned with V? thank you

**Comment by JalonSolov** [comment]
No.  V will never be object oriented.  It does have some similar functionality (such as being able to add methods to structs), but it doesn't have inheritance, etc.

**Comment by Wajinn** [comment]
> Hi, I am a newcomer. I dont see any mention of Object oriented. Is object oriented planned with V?

V can be considered as object oriented, but it depends on what definitions people believe qualifies.  Such debates can be very subjective.  The definition of OOP, as defined by Alan Kay, is not what is now known as class-based OOP (which exists in Java and C++).  Many people also don't realize that there are other versions, besides class-based, such as prototype-based OOP (as seen in JavaScript and Lua). 

V arguably subscribes to a more generalized version of OOP, similar to that of Golang.  Which is not too far removed to how OOP can be done in JavaScript, Lua, or hybrid languages such as Object Pascal, where classes aren't or don't have to be used.  This is using objects/structs/records, that can be embeddable and be assigned methods.  In the majority of situations, this generalized version of OOP, can be more preferable and do what is done with class-based versions (though some creativity might be required).  Additionally, Vlang has features such as closures, which (combined with other features) greatly minimizes or negates the need for class-based OOP.

There are also many who call into question the necessity of class-based OOP, where its usage is forced as primary, and can be seen as something possibly needed 5% of the time or less, only needed in rare specific circumstances, or not needed at all.  Often composition is considered more useful than inheritance.  Encapsulation can be done more generally through the use of modules or units, that various languages have.  Polymorphism is not specific to only OOP, and various forms of it can be used with other paradigms.

**Comment by ghost** [comment]
The term object-oriented has many definitions.
Some consider Go and Rust to be object-oriented languages, while others do not.
The author of V does not consider V an object-oriented language.
In fact, like Go, it has no inheritance and its encapsulation is incomplete.
It can be considered almost identical to Go.

**Comment by timi5691** [comment]
what a bout entity component system library in v?

**Reply by JalonSolov** [reply]
You mean like for video games?  There are a few people who have talked about video games in V on discord in the #gamedev channel.
